test/mocha/20-tokens.js

```diff
@@ -1,18 +1,398 @@
+      const {tokens: tks} = await tokens.create(
+        {internalId, attributes, tokenCount});
+      const token = tks[0];
+      result = await tokens.resolveToInternalId({token});
+    } catch(e) {
+      err = e;
+    }
+    should.not.exist(err);
+    should.exist(result);
+    result.should.be.an('object');
+    result.should.deep.equal({internalId: 'aM6pup9s6XTaYTho'});
+  });
+  it('should throw error when wrapped value fails to get decrypted',
+    async function() {
+      const tokenCount = 1;
+      const attributes = Uint8Array.from(new Set([1]));
+      const internalId = 'aM6pup9s6XTaYTho';
+      const requester = 'requester';
+      let err;
+      let result;
+      try {
+        const {tokens: tks} = await tokens.create(
+          {internalId, attributes, tokenCount});
+        const token = tks[0];
+        // change wrapped value by altering its first index
+        token[26] = 0;
+        result = await tokens.resolve({requester, token});
+      } catch(e) {
+        err = e;
+      }
+      should.not.exist(result);
+      should.exist(err);
+      err.name.should.equal('DataError');
+      err.message.should.equal('Invalid token.');
+      err.cause.message.should.equal('Decryption failed.');
+    });
+  it('should throw error when attributes is incorrect.',
+    async function() {
+      const tokenCount = 1;
+      const attributes = Uint8Array.from(new Set([1]));
+      const internalId = 'aM6pup9s6XTaYTho';
+      const requester = 'requester';
+      let err;
+      let result;
+      try {
+        const {tokens: tks} = await tokens.create(
+          {internalId, attributes, tokenCount});
+        const token = tks[0];
+        // change attributes value by altering its index at 50
+        token[50] = 0;
+        result = await tokens.resolve({requester, token});
+      } catch(e) {
+        err = e;
+      }
+      should.not.exist(result);
+      should.exist(err);
+      err.name.should.equal('DataError');
+      err.message.should.equal('Invalid token.');
+    });
+});
+
+describe('TokensDuplicateError', function() {
+  let randomBytesStub;
+  before(() => {
+    randomBytesStub = sinon.stub(crypto, 'randomBytes').callsFake(
+      (size, cb) => {
+        if(size > MAX_UINT32) {
+          throw new RangeError('requested too many random bytes');
+        }
+        const bytes = Buffer.alloc(size);
+        if(typeof cb === 'function') {
+          return process.nextTick(function() {
+            cb(null, bytes);
+          });
+        }
+        return bytes;
+      });
+  });
+  after(() => {
+    randomBytesStub.restore();
+  });
+  it('should throw duplicate error if token is created twice.',
+    async function() {
+      const tokenCount = 100;
+      const attributes = Uint8Array.from(new Set([1]));
+      const internalId = 'aM6pup9s6XTaYTho';
+      let err;
+      let result1;
+      let result2;
+      try {
+        result1 = await tokens.create({internalId, attributes, tokenCount});
```

Comment:
comments are needed here that explain that the setup here is that:
`crypto.randomBytes` has been stubbed so that the same batch ID will be generated every time, and we are consuming all the tokens in the first batch associated with the internalId so that when a new batch is created, the same ID is generated and the error occurs.